libraries/Microsoft.Bot.Builder.Runtime/Microsoft.Bot.Builder.Runtime.csproj

```diff
@@ -0,0 +1,63 @@
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs.Adaptive" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs.Debugging" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs.Debugging" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+
+    <PackageReference Include="Microsoft.Bot.Builder.AI.Luis" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.Luis" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.ApplicationInsights" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.ApplicationInsights" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.ApplicationInsights.Core" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.ApplicationInsights.Core" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Azure" Condition=" '$(ReleasePackageVersion)' == '' " Version="$(LocalPackageVersion)" />
+    <PackageReference Include="Microsoft.Bot.Builder.Azure" Condition=" '$(ReleasePackageVersion)' != '' " Version="$(ReleasePackageVersion)" />  
+
+    <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.8" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.Bot.Builder.Dialogs.Adaptive\Microsoft.Bot.Builder.Dialogs.Adaptive.csproj" />
+    <ProjectReference Include="..\Microsoft.Bot.Builder.Dialogs.Debugging\Microsoft.Bot.Builder.Dialogs.Debugging.csproj" />
+    <ProjectReference Include="..\Microsoft.Bot.Builder.Dialogs.Adaptive\Microsoft.Bot.Builder.Dialogs.Adaptive.csproj" />
+    <ProjectReference Include="..\integration\Microsoft.Bot.Builder.Integration.AspNet.Core\Microsoft.Bot.Builder.Integration.AspNet.Core.csproj" />
```

Comment:
Since this project is AspNet.Core specific, it may be better to move it to the integration folder (where all the other specific projects live).

Reply:
I can see why you'd suggest this.  Though the difference is that the other integrations provide code to complete functionality in another package.  As of this writing, I don't see there would be a WebApi integration since that is missing (at least) SkillHttpClient.  Apparently we don't support Skills on WebApi?

Reply:
@gabog providing response to your questions outlined above:

> First pass at reviewing this.
> 
> I checked general project structure and dependencies at this point. I'd like to understand what Builders and Providers do and what will the extensibility points are before reviewing the rest.
> 
> Some thoughts include:
> 
> 1. Will the dev be able to change Azure services by something else?
> 2. Will the dev be able to change QnA and or LUIS by something else?
> 3. How will the dev register other cognitive services? (add image recognizer for example)
> 4. Will the dev be able to change the adapter and use a different one?
> 5. Will we let the dev change or swap telemetry?
> 6. Will they be able to use TeamsActivityHandler instance of ActivityHandler?

Our primary design aim with this runtime code is to enable modifying the behaviors of the runtime without needing to make modifications to the code. To that end, we are leveraging a declarative JSON file to describe the runtime behaviors utilizing the Adaptive library, and this is where we can modify elements such as which adapter to use, storage service, middleware pipeline, etc.

Right now, we have incorporated this into a separate 'runtime.json' file that lives in the root directory of the project, but we are going to evaluate collapsing this into the .botproj file so as to minimize additional files to the project. Currently, it looks something like this:

`{
    "adapters": [
    {
        "$kind": "Microsoft.BotCoreAdapter",
        "middleware": [
        {
            "$kind": "Microsoft.TranscriptLoggerMiddleware",
            "transcriptStore": {
                "$kind": "Microsoft.BlobsTranscriptStore",
                "connectionString": "=blobStorage.connectionString",
                "containerName": "=blobStorage.container"
            }
        },
        {
            "$kind": "Microsoft.ShowTypingMiddleware",
            "delay": 500,
            "period": 2000
        }]
    }],
    "channel": {
        "$kind": "Microsoft.DeclarativeChannelProvider",
        "channel": null
    },
    "credentials": {
        "$kind": "Microsoft.DeclarativeCredentialsProvider",
        "applicationId": "=MicrosoftAppId",
        "applicationPassword": "=MicrosoftAppPassword"
    },
    "defaultLocale": "en-US",
    "removeRecipientMention": true,
    "rootDialog": "=bot",
    "storage": {
        "$kind": "Microsoft.CosmosDbPartitionedStorage",
        "authenticationKey": "=cosmosDb.authKey",
        "compatibilityMode": true,
        "containerId": "=cosmosDb.containerId",
        "containerThroughput": 400,
        "databaseId": "=cosmosDb.databaseId",
        "endpoint": "=cosmosDb.cosmosDBEndpoint",
        "keySuffix": null
    },
    "telemetry": {
        "$kind": "Microsoft.ApplicationInsightsTelemetryProvider",
        "instrumentationKey": "=applicationInsights.InstrumentationKey"
    }
}`

This JSON is loaded through the ResourceExplorer class and utilizes the same type bindings that we use for our other Adaptive pieces, so it allows us to add additional types through registering schemas in the same way (I'll touch on this more below).

We want to also abide by the established pattern of registering services with IServiceCollection and making them available for DI/IoC using the standard mechanisms. To that end, we utilize Providers as an implementation against the different schema types as a class which is responsible for registering services to be utilized through DI. However, in some cases (e.g. middleware, these need to really be instantiated within the scope of the injected service. For that case, Builders provide a common interface which, given access to registered services and app configuration, produces an instance of a particular class.

To that end, should users wish to add custom providers for things like storage or adapters, etc., then it is a simple case of providing the type registration to the Adaptive pipeline and just updating the JSON accordingly. This pattern can then be utilized to change things like telemetry, adapters, etc. For now, we have a single standard activity handler implementation we are using, but this is something that we can look at changing as necessary.

---

One area that is not yet present but we are beginning to look into is our overall extensibility model, e.g. if I want to use Orchestrator as a recognizer that is defined in an external package, how do I incorporate it here?

At a high level, the current thinking is that users will add additional package references containing these definitions, and through some mechanism we will discover these types through the runtime library and perform the necessary registration of those components such that they can fit into the same structure. That way, as long as the necessary package references are present, then the only thing the user should need to do to modify is update the appropriate declarative assets (e.g. update runtime.json to use a different provider, update .dialog to use a custom recognizer, etc).

We have a discussion planned today with some key folks across the SDK team with expertise in different languages to make sure that the high-level approach is feasible within those language ecosystems, but once we've made some headway there, we'll start working on incorporating those capabilities into here.

Let me know if you have any questions in the meantime and I'll be happy to answer them!

Reply:
Sorry, GitHub replies are weird - my above is really more a general response to the review message, not so much the specific one above.

To that point though, you're right in that we should probably adjust taking dependencies ideally on specific integration packages. It could be that there is overlap, so I think we need to pull at the threads and untangle where necessary. I'm not sure this is going to be a simple thing to do right this moment, so we can cut an issue to track evaluation and handling this.